apps/site/pages/en/learn/diagnostics/memory/understanding-and-tuning-memory.md

```diff
@@ -0,0 +1,137 @@
+
+### The Stack
+
+In addition to the heap, V8 also uses the **stack** for memory management. The stack is a region of memory used to store local variables and function call information. Unlike the heap, which is managed by V8's garbage collector, the stack operates on a Last In, First Out (LIFO) principle.
+
+Whenever a function is called, a new frame is pushed onto the stack. When the function returns, its frame is popped off. The stack is much smaller in size compared to the heap, but it is faster for memory allocation and deallocation. However, the stack has a limited size, and excessive use of memory (such as with deep recursion) can result in a **stack overflow**.
+
+## Monitoring Memory Usage
+
+Before tuning memory usage, it’s important to understand how much memory your application is consuming. Node.js and V8 provide several tools for monitoring memory usage.
+
+### Using `process.memoryUsage()`
+
+The `process.memoryUsage()` method provides insights into how much memory your Node.js process is using. It returns an object with details like:
+
+- **`rss`** (Resident Set Size): The total memory allocated to your process, including heap and other areas.
+- **`heapTotal`**: The total memory allocated for the heap.
+- **`heapUsed`**: The memory currently in use within the heap.
+- **`external`**: Memory used by external resources like bindings to C++ libraries.
+- **`arrayBuffers`**: Memory allocated to various Buffer-like objects.
+
+Here’s an example:
+
+```javascript
+console.log(process.memoryUsage());
+```
+
+The output will look like:
+
+```json
+{
+  "rss": 25837568,
+  "heapTotal": 5238784,
+  "heapUsed": 3666120,
+  "external": 1274076,
+  "arrayBuffers": 10515
+}
+```
+
+By monitoring these values over time, you can identify if memory usage continually increases over time, a common sign of memory leaks.
+
+## Command-Line Flags for Memory Tuning
+
+Node.js offers several command-line flags that help you fine-tune memory-related settings. Below are some key options you can use to optimize memory usage.
+
+### `--max-old-space-size`
+
+This flag sets a limit on the size of the **Old Space** in the V8 heap, where long-lived objects are stored. If your application uses a significant amount of memory, you might need to adjust this limit.
+
+To increase the Old Space limit to 4 GB, for example, use:
+
+```bash
+node --max-old-space-size=4096 app.js
+```
+
+This sets the Old Space size to 4096 MB (4 GB), but you should adjust this based on your system's available memory.
+
+### `--max-semi-space-size`
+
+The `--max-semi-space-size` flag controls the size of the **New Space** in the V8 heap. New Space stores newly created objects, which are garbage collected frequently. Increasing this size can help reduce the frequency of minor garbage collections, which can in turn improve performance if your application allocates a lot of short-lived objects.
+
+For example:
+
+```bash
+node --max-semi-space-size=512 app.js
```

Comment:
I would not recommend this to be 512. 64 is usually the best compromise between memory consumption and performance.

Note that the default is based on old space size, and in some limited cases is 1 MB, making it _hard_ to not thrash the garbage collector.